pkg/controller/istiocontrolplane/istiocontrolplane_controller_test.go

```diff
@@ -0,0 +1,196 @@
+		},
+		Spec: &v1alpha2.IstioControlPlaneSpec{
+			Profile: "minimal",
+		},
+	}
+	objs := []runtime.Object{
+		icp,
+	}
+
+	s := scheme.Scheme
+	s.AddKnownTypes(v1alpha2.SchemeGroupVersion, icp)
+	cl := fake.NewFakeClientWithScheme(s, objs...)
+	factory := &helmreconciler.Factory{CustomizerFactory: &IstioRenderingCustomizerFactory{}}
+	r := &ReconcileIstioControlPlane{client: cl, scheme: s, factory: factory}
+
+	req := reconcile.Request{
+		NamespacedName: types.NamespacedName{
+			Name:      name,
+			Namespace: namespace,
+		},
+	}
+	res, err := r.Reconcile(req)
+	if err != nil {
+		t.Fatalf("reconcile: (%v)", err)
+	}
+	// check ICP status
+	succeed, err := checkICPStatus(cl, req.NamespacedName, "minimal")
+	if !succeed || err != nil {
+		t.Fatalf("failed to get expected IstioControlPlane status: (%v)", err)
+	}
+
+	//update IstioControlPlane : switch profile from minimal to default and reconcile
+	err = switchIstioControlPlaneProfile(cl, req.NamespacedName, "default")
+	if err != nil {
+		t.Fatalf("failed to update IstioControlPlane: (%v)", err)
+	}
+	res, err = r.Reconcile(req)
+	if err != nil {
+		t.Fatalf("reconcile: (%v)", err)
+	}
+	if res.Requeue {
+		t.Error("reconcile requeue which is not expected")
+	}
+	// check ICP status
+	succeed, err = checkICPStatus(cl, req.NamespacedName, "default")
+	if !succeed || err != nil {
+		t.Fatalf("failed to get expected IstioControlPlane status: (%v)", err)
+	}
+
+	//update IstioControlPlane : switch profile from default to minimal again and reconcile
+	err = switchIstioControlPlaneProfile(cl, req.NamespacedName, "minimal")
+	if err != nil {
+		t.Fatalf("failed to update IstioControlPlane: (%v)", err)
+	}
+	res, err = r.Reconcile(req)
+	if err != nil {
+		t.Fatalf("reconcile: (%v)", err)
+	}
+	// Check the result of reconciliation to make sure it has the desired state.
+	if res.Requeue {
+		t.Error("reconcile requeue which is not expected")
+	}
+	// check ICP status
+	succeed, err = checkICPStatus(cl, req.NamespacedName, "minimal")
+	if !succeed || err != nil {
+		t.Fatalf("failed to get expected IstioControlPlane status: (%v)", err)
+	}
+
+}
+func statusExpected(s1, s2 *v1alpha2.InstallStatus_VersionStatus) bool {
+	return s1.Status.String() == s2.Status.String()
+}
+
+func switchIstioControlPlaneProfile(cl client.Client, key client.ObjectKey, profile string) error {
+	instance := &v1alpha2.IstioControlPlane{}
+	err := cl.Get(context.TODO(), key, instance)
+	if err != nil {
+		return err
+	}
+	instance.Spec.Profile = profile
+	err = cl.Update(context.TODO(), instance)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+func checkICPStatus(cl client.Client, key client.ObjectKey, profile string) (bool, error) {
+	minimalStatus := map[string]*v1alpha2.InstallStatus_VersionStatus{
+		"Pilot": {
+			Status: v1alpha2.InstallStatus_HEALTHY,
+		},
+		"crds": {
+			Status: v1alpha2.InstallStatus_HEALTHY,
+		},
+	}
+	defaultStatus := map[string]*v1alpha2.InstallStatus_VersionStatus{
```

Comment:
Maybe we can make these variables be constant in follow up PR as we need to more profile switch cases.